plugin/db/util/driverutil.go

```diff
@@ -201,10 +201,23 @@ func beginMigration(ctx context.Context, executor MigrationExecutor, m *db.Migra
 
 	// Phase 1 - Precheck before executing migration
 	// Check if the same migration version has alraedy been applied
-	if duplicate, err := executor.CheckDuplicateVersion(ctx, tx, m.Namespace, m.Source, m.Version); err != nil {
-		return -1, err
-	} else if duplicate {
-		return -1, common.Errorf(common.MigrationAlreadyApplied, fmt.Errorf("database %q has already applied version %s", m.Database, m.Version))
+	list, err := executor.FindMigrationHistoryList(ctx, &db.MigrationHistoryFind{
+		Database: &m.Namespace,
+		Source:   &m.Source,
+		Version:  &m.Version,
+	})
+	if err != nil {
+		return -1, fmt.Errorf("Check duplicate version error: %q", err)
+	} else if len(list) > 0 {
+		switch list[0].Status {
+		case db.Done:
+			return -1, common.Errorf(common.MigrationAlreadyApplied, fmt.Errorf("database %q has already applied version %s", m.Database, m.Version))
+		case db.Pending:
+			return -1, common.Errorf(common.MigrationAlreadyApplied, fmt.Errorf("database %q version %s is pending", m.Database, m.Version))
+		case db.Failed:
```

Comment:
I am not that confident to retry if failed

If we look at ExecuteMigration, there exist different failure modes, and maybe the database is in an inconsistent state. I would rather abort the migration and let DBA / Developer manually inspect them instead of continuing the operation which could worsen the situation.

I am OK to return different error code for Done/Pending/Failed state respectively.

Reply:
Yes, we can start with better error messages first.

I think it's invoked when someone approves the task by changing task from FAILED to PENDING status, assuming someone has already inspected the database already. The error can happen when any step in ExecuteMigration() failed. More importantly, when we support gh-ost (long-time to migrate), it's more likely to fail in the middle. I'm thinking of adding more information about FAILED status to indicate whether executor.Execute() has been completed or not. If not, we can do the logic in this PR. Otherwise, we should continue the dump step after the Execute().

https://github.com/bytebase/bytebase/blob/main/plugin/db/util/driverutil.go#L169